src/engine/fcitx/openbangla.cpp

```diff
@@ -0,0 +1,388 @@
+    case FcitxKey_ISO_Level3_Shift:
+      // Keep track of the right Alt key (also known as the AltGr key)
+      altGrPressed_ = true;
+      [[fallthrough]];
+    case FcitxKey_Shift_L:
+    case FcitxKey_Shift_R:
+    case FcitxKey_Control_L:
+    case FcitxKey_Control_R:
+    case FcitxKey_Alt_L:
+    case FcitxKey_Meta_L:
+    case FcitxKey_Meta_R:
+      if (riti_context_ongoing_input_session(ctx)) {
+        keyEvent.filterAndAccept();
+      }
+      return;
+    default:
+      break;
+    }
+
+    uint8_t modifier = 0;
+    // Set modifiers
+    if (key.states().test(KeyState::Shift)) {
+      modifier |= MODIFIER_SHIFT;
+    }
+
+    auto ctrlKey = key.states().test(KeyState::Ctrl);
+    auto altKey = key.states().test(KeyState::Alt);
+
+    // Convert the key value into riti's key value.
+    auto ritiKey = keySymToRitiKey(key.sym());
+
+    // Reject the key which has only Ctrl or Alt (not the right one) combination
+    // and riti doesn't handle.
+    if ((ctrlKey && !altKey) || (!ctrlKey && altKey && !altGrPressed_) ||
+        ritiKey == VC_UNKNOWN) {
+      if (riti_context_ongoing_input_session(ctx)) {
+        commit();
+      }
+      return;
+    }
+
+    // If we have Ctrl and Alt combination or the right Alt, set it as the AltGr
+    // modifier.
+    if ((ctrlKey && altKey) || altGrPressed_) {
+      modifier |= MODIFIER_ALT_GR;
+    }
+
+    suggestion_.reset(riti_get_suggestion_for_key(ctx, ritiKey, modifier));
+
+    if (!riti_suggestion_is_empty(suggestion_.get())) {
+      updateUI();
+      keyEvent.filterAndAccept();
+    }
+  }
+
+private:
+  OpenBanglaEngine *engine_;
+  InputContext *ic_;
+  /* Unfortunately, we have to keep track of the right Alt Key. */
+  bool altGrPressed_ = false;
+  UniqueCPtr<RitiContext, riti_context_free> ctx_;
+  UniqueCPtr<Suggestion, riti_suggestion_free> suggestion_;
+};
+
+OpenBanglaEngine::OpenBanglaEngine(Instance *instance)
+    : instance_(instance), factory_([this](InputContext &ic) {
+        return new OpenBanglaState(this, ic);
+      }) {
+  if (!fs::makePath(stringutils::joinPath(
+          StandardPath::global().userDirectory(StandardPath::Type::Data),
+          "openbangla-keyboard"))) {
+    throw std::runtime_error("Failed to create user directory");
+  }
+  reloadConfig();
+  instance->inputContextManager().registerProperty("openbanglaState",
+                                                   &factory_);
+}
+
+OpenBanglaEngine::~OpenBanglaEngine() {}
+
+void OpenBanglaEngine::keyEvent(const InputMethodEntry &, KeyEvent &keyEvent) {
+  auto ic = keyEvent.inputContext();
+  auto state = ic->propertyFor(&factory_);
+  state->keyEvent(keyEvent);
+}
+
+void OpenBanglaEngine::reset(const InputMethodEntry &,
+                             InputContextEvent &event) {
+  auto state = event.inputContext()->propertyFor(&factory_);
+  state->reset();
+}
+
+void OpenBanglaEngine::populateConfig() {
+  setenv("RITI_LAYOUT_FILE", config_.layoutPath->data(), 1);
```

Comment:
We have to use the [Settings](https://github.com/OpenBangla/OpenBangla-Keyboard/blob/master/src/shared/Settings.h) class to get the configurations. The configurations have to be shared between the GUI(Qt frontend) and this plugin.

Reply:
I understand you want to have a shared settings file for everything, but link Qt to plugin in a non-Qt application is a kinda dangerous thing to do. (mainly related to dlopen mechanism)

I would avoid to use QSettings in any fcitx code, but I could use fcitx's ini parser to read the same settings file if that would work for you. Also using Qt API without application is likely to have some pitfalls so I would try to avoid that too.

As for monitor config file change, I'd suggest
1. either use a dbus signal to notify any changes from frontend, so we don't have to unnecessarily re-read file everywhere. That's a more ideal situation.
2. or fcitx would use some filesystem monitoring (e.g. inotify) on the settings file.

Reply:
> I understand you want to have a shared settings file for everything, but link Qt to plugin in a non-Qt application is a kinda dangerous thing to do. (mainly related to dlopen mechanism)

Yeah, I also agree.
>  but I could use fcitx's ini parser to read the same settings file if that would work for you.

I am fine with that :smile: 

> As for monitor config file change, I'd suggest

I also agree that the ideal solution is to use the dbus signal mechanism, but this would be a big change to handle within this PR. I will be really eager to handle this stuff with you in the next PR's if you want, that would be amazing.

For now, I think you can use the inotify mechanism or that ugly hack I am using in ibus engine :sweat_smile: for this initial PR.

Sorry for being late! :sweat: